Double promise mark warning in case of bad certificate

If caller provides "bad" certificate to Pushy (i.e. certificate that causes CERTIFICATE_AND_PRIVATE_KEY_MISMATCH or SSLV3_ALERT_CERTIFICATE_REVOKED SSL error) pushy logs following warning:

> WARN  [apns-push-6] [2019-04-11T12:07:55.911Z] [c.t.p.a.ApnsChannelFactory] [] [] [] [] Tried to mark promise as "failed," but it was already done.

It seems to be a design problem, but more important that service logs are filled by warnings that look to be pointless (this problem have to be resolved and/or logged by caller)

I'm currently dealing with this issue and it seems to be bit complicated.
Software versions: 
Pushy 0.14
Netty 4.1.34.Final (I've experienced this issue with previous versions as well
netty-tcnative-boringssl-static 2.0.25.Final
O/S: Amazon Linux AMI 2017.09 and Ubuntu Linux 18.04
Native *ssl* packages openssl-1.0.2k-8.107.amzn1.x86_64 and 1.1.0g-2ubuntu4.3 respectively. I have no ideas whether they are used.

Client provides bad certificate+key. One can reproduce this by providing key for different certificate, or by providing key+cert that are already revoked in Apple console.

Every time this causes the following warning:

> 
> WARN  [apns-push-1] [2019-04-18T10:45:23.529Z] [c.t.p.a.ApnsChannelFactory] [] [] [] [] Tried to mark promise as "failed," but it was already done.
> io.netty.handler.codec.DecoderException: javax.net.ssl.SSLHandshakeException: error:10000414:SSL routines:OPENSSL_internal:SSLV3_ALERT_CERTIFICATE_REVOKED
> 	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:472)
> 	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:278)
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:345)
> 	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:337)
> 	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1408)
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:345)
> 	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:930)
> 	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
> 	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:677)
> 	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:612)
> 	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:529)
> 	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:491)
> 	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:905)
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
> 	at java.lang.Thread.run(Thread.java:748)
> Caused by: javax.net.ssl.SSLHandshakeException: error:10000414:SSL routines:OPENSSL_internal:SSLV3_ALERT_CERTIFICATE_REVOKED
> 	at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.shutdownWithError(ReferenceCountedOpenSslEngine.java:973)
> 	at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.sslReadErrorResult(ReferenceCountedOpenSslEngine.java:1231)
> 	at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.unwrap(ReferenceCountedOpenSslEngine.java:1185)
> 	at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.unwrap(ReferenceCountedOpenSslEngine.java:1256)
> 	at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.unwrap(ReferenceCountedOpenSslEngine.java:1299)
> 	at io.netty.handler.ssl.SslHandler$SslEngineType$1.unwrap(SslHandler.java:217)
> 	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1330)
> 	at io.netty.handler.ssl.SslHandler.decodeNonJdkCompatible(SslHandler.java:1237)
> 	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1274)
> 	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:502)
> 	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:441)
> 	... 17 common frames omitted

I've added some more logic to test what's going on (changed that error message to provide more details
and added print stack trace):
> Tried to mark promise as "failed," with cause='io.netty.handler.codec.DecoderException: javax.net.ssl.SSLHandshakeException: error:10000414:SSL routines:OPENSSL_internal:SSLV3_ALERT_CERTIFICATE_REVOKED', but it is already 'Failed' with cause='javax.net.ssl.SSLHandshakeException: error:10000414:SSL routines:OPENSSL_internal:SSLV3_ALERT_CERTIFICATE_REVOKED'

First call:

> java.lang.Exception: Stack trace
> 	at java.lang.Thread.dumpStack(Thread.java:1336)
> 	at com.turo.pushy.apns.ApnsChannelFactory.tryFailureAndLogRejectedCause(ApnsChannelFactory.java:264)
> 	at com.turo.pushy.apns.ApnsChannelFactory.access$100(ApnsChannelFactory.java:56)
> 	at com.turo.pushy.apns.ApnsChannelFactory$1$1.operationComplete(ApnsChannelFactory.java:162)
> 	at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:502)
> 	at io.netty.util.concurrent.DefaultPromise.notifyListeners0(DefaultPromise.java:495)
> 	at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:474)
> 	at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:415)
> 	at io.netty.util.concurrent.DefaultPromise.setValue0(DefaultPromise.java:540)
> 	at io.netty.util.concurrent.DefaultPromise.setFailure0(DefaultPromise.java:533)
> 	at io.netty.util.concurrent.DefaultPromise.tryFailure(DefaultPromise.java:114)
> 	at io.netty.handler.ssl.SslHandler.handleUnwrapThrowable(SslHandler.java:1249)
> 	at io.netty.handler.ssl.SslHandler.decodeNonJdkCompatible(SslHandler.java:1239)
> 	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1274)
> 	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:502)
> 	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:441)
> 	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:278)
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:345)
> 	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:337)
> 	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1408)
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:345)
> 	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:930)
> 	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
> 	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:677)
> 	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:612)
> 	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:529)
> 	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:491)
> 	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:905)
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
> 	at java.lang.Thread.run(Thread.java:748)

This means that tryFailureAndLogRejectedCause was invoked in last "else" of listener "sslHandler.handshakeFuture().addListener(new GenericFutureListener<Future<Channel>>() "

Second call: 

> java.lang.Exception: Stack trace
> 	at java.lang.Thread.dumpStack(Thread.java:1336)
> 	at com.turo.pushy.apns.ApnsChannelFactory.tryFailureAndLogRejectedCause(ApnsChannelFactory.java:264)
> 	at com.turo.pushy.apns.ApnsChannelFactory.access$100(ApnsChannelFactory.java:56)
> 	at com.turo.pushy.apns.ApnsChannelFactory$ConnectionNegotiationErrorHandler.exceptionCaught(ApnsChannelFactory.java:80)
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:282)
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:261)
> 	at io.netty.channel.AbstractChannelHandlerContext.fireExceptionCaught(AbstractChannelHandlerContext.java:253)
> 	at io.netty.handler.ssl.SslHandler.exceptionCaught(SslHandler.java:1096)
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:282)
> 	at io.netty.channel.AbstractChannelHandlerContext.notifyHandlerException(AbstractChannelHandlerContext.java:820)
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:361)
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:345)
> 	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:337)
> 	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1408)
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
> 	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:345)
> 	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:930)
> 	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
> 	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:677)
> 	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:612)
> 	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:529)
> 	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:491)
> 	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:905)
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
> 	at java.lang.Thread.run(Thread.java:748)
It means method was invoked by ConnectionNegotiationErrorHandler::exceptionCaught

But basically this warning is a half of evil. Sometimes with same condition (same certificates) I experience much serious issue with SSL. I can't reproduce this error and it appears only several times per day on AWS box.  Some error appears at SSL level, HTTP/2 connection is dropped during preface and following errors are logged:

One:
> [sendNotification] PushFailure, status='FAIL' apnsError='None' message='Failed to dispatch message to APNS, reason: Channel closed before HTTP/2 preface completed.' apnsId='e5c38271-ffc4-25ad-766e-34a6b11b96ae' cause='java.lang.IllegalStateException', cause.message='Channel closed before HTTP/2 preface completed.'
> It means that in ApnsChannelFactory:229 the connectFuture.channel().closeFuture().addListener(new GenericFutureListener<ChannelFuture> ()::operationComplete is invoked with IllegalStateException 
> 


Two:

> WARN  [apns-push-5] [2019-04-17T21:22:03.162Z] [i.n.c.ChannelInitializer] [] [] [] [] Failed to initialize a channel. Closing: [id: 0xdfc9280f]
> java.lang.NullPointerException: ctx
>         at io.netty.internal.tcnative.SSL.newSSL(Native Method)
>         at io.netty.handler.ssl.ReferenceCountedOpenSslEngine.<init>(ReferenceCountedOpenSslEngine.java:338)
>         at io.netty.handler.ssl.ReferenceCountedOpenSslContext.newEngine0(ReferenceCountedOpenSslContext.java:419)
>         at io.netty.handler.ssl.ReferenceCountedOpenSslContext.newHandler(ReferenceCountedOpenSslContext.java:399)
>         at io.netty.handler.ssl.SslContext.newHandler(SslContext.java:888)
>         at com.turo.pushy.apns.ApnsChannelFactory$1.initChannel(ApnsChannelFactory.java:119)
>         at com.turo.pushy.apns.ApnsChannelFactory$1.initChannel(ApnsChannelFactory.java:109)
>         at io.netty.channel.ChannelInitializer.initChannel(ChannelInitializer.java:129)
>         at io.netty.channel.ChannelInitializer.handlerAdded(ChannelInitializer.java:112)
>         at io.netty.channel.AbstractChannelHandlerContext.callHandlerAdded(AbstractChannelHandlerContext.java:953)
>         at io.netty.channel.DefaultChannelPipeline.callHandlerAdded0(DefaultChannelPipeline.java:610)
>         at io.netty.channel.DefaultChannelPipeline.access$100(DefaultChannelPipeline.java:46)
>         at io.netty.channel.DefaultChannelPipeline$PendingHandlerAddedTask.execute(DefaultChannelPipeline.java:1461)
>         at io.netty.channel.DefaultChannelPipeline.callHandlerAddedForAllHandlers(DefaultChannelPipeline.java:1126)
>         at io.netty.channel.DefaultChannelPipeline.invokeHandlerAddedIfNeeded(DefaultChannelPipeline.java:651)
>         at io.netty.channel.AbstractChannel$AbstractUnsafe.register0(AbstractChannel.java:515)
>         at io.netty.channel.AbstractChannel$AbstractUnsafe.access$200(AbstractChannel.java:428)
>         at io.netty.channel.AbstractChannel$AbstractUnsafe$1.run(AbstractChannel.java:487)
>         at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
>         at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:404)
>         at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:495)
>         at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:905)
>         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
>         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
>         at java.lang.Thread.run(Thread.java:748)
> INFO  [apns-push-0] [2019-04-17T21:22:03.162Z] [c.t.r.a.v.s.TwilioApnsClient] [RQ6e428e46b9cf41d0bc028b2ef36bde53] [RU5c797e6006dba154e874b4a9b6ee986b] [ACf5eaf92c0dbeb386aed1e01b353bd92e] [sendNotification] sendNotification: message='APNs future completed' latency='1064'

It looks like some problem with native SSL.

--- a/pushy/src/main/java/com/eatthepath/pushy/apns/ApnsChannelFactory.java
+++ b/pushy/src/main/java/com/eatthepath/pushy/apns/ApnsChannelFactory.java
@@ -75,17 +75,6 @@ class ApnsChannelFactory implements PooledObjectFactory<Channel>, Closeable {
 
     private static final Logger log = LoggerFactory.getLogger(ApnsChannelFactory.class);
 
-    @ChannelHandler.Sharable
-    private static class ConnectionNegotiationErrorHandler extends ChannelInboundHandlerAdapter {
-
-        static final ConnectionNegotiationErrorHandler INSTANCE = new ConnectionNegotiationErrorHandler();
-
-        @Override
-        public void exceptionCaught(final ChannelHandlerContext context, final Throwable cause) {
-            tryFailureAndLogRejectedCause(context.channel().attr(CHANNEL_READY_PROMISE_ATTRIBUTE_KEY).get(), cause);
-        }
-    }
-
     ApnsChannelFactory(final SslContext sslContext, final ApnsSigningKey signingKey, final long tokenExpirationMillis,
                        final ProxyHandlerFactory proxyHandlerFactory, final int connectTimeoutMillis,
                        final long idlePingIntervalMillis, final long gracefulShutdownTimeoutMillis,
@@ -116,49 +105,12 @@ class ApnsChannelFactory implements PooledObjectFactory<Channel>, Closeable {
 
             @Override
             protected void initChannel(final SocketChannel channel) {
-                final ChannelPipeline pipeline = channel.pipeline();
-
-                if (proxyHandlerFactory != null) {
-                    pipeline.addFirst(proxyHandlerFactory.createProxyHandler());
-                }
-
                 final SslHandler sslHandler = sslContext.newHandler(channel.alloc());
 
                 sslHandler.handshakeFuture().addListener(new GenericFutureListener<Future<Channel>>() {
                     @Override
                     public void operationComplete(final Future<Channel> handshakeFuture) {
                         if (handshakeFuture.isSuccess()) {
-                            final String authority = channel.remoteAddress().getHostName();
-
-                            final ApnsClientHandler.ApnsClientHandlerBuilder clientHandlerBuilder;
-
-                            if (signingKey != null) {
-                                clientHandlerBuilder = new TokenAuthenticationApnsClientHandler.TokenAuthenticationApnsClientHandlerBuilder()
-                                        .signingKey(signingKey)
-                                        .tokenExpirationMillis(tokenExpirationMillis)
-                                        .authority(authority)
-                                        .idlePingIntervalMillis(idlePingIntervalMillis);
-                            } else {
-                                clientHandlerBuilder = new ApnsClientHandler.ApnsClientHandlerBuilder()
-                                        .authority(authority)
-                                        .idlePingIntervalMillis(idlePingIntervalMillis);
-                            }
-
-                            if (frameLogger != null) {
-                                clientHandlerBuilder.frameLogger(frameLogger);
-                            }
-
-                            final ApnsClientHandler apnsClientHandler = clientHandlerBuilder.build();
-
-                            if (gracefulShutdownTimeoutMillis > 0) {
-                                apnsClientHandler.gracefulShutdownTimeoutMillis(gracefulShutdownTimeoutMillis);
-                            }
-
-                            pipeline.addLast(new FlushConsolidationHandler(FlushConsolidationHandler.DEFAULT_EXPLICIT_FLUSH_AFTER_FLUSHES, true));
-                            pipeline.addLast(new IdleStateHandler(idlePingIntervalMillis, 0, 0, TimeUnit.MILLISECONDS));
-                            pipeline.addLast(apnsClientHandler);
-                            pipeline.remove(ConnectionNegotiationErrorHandler.INSTANCE);
-
                             channel.attr(CHANNEL_READY_PROMISE_ATTRIBUTE_KEY).get().trySuccess(channel);
                         } else {
                             tryFailureAndLogRejectedCause(channel.attr(CHANNEL_READY_PROMISE_ATTRIBUTE_KEY).get(), handshakeFuture.cause());
@@ -166,8 +118,45 @@ class ApnsChannelFactory implements PooledObjectFactory<Channel>, Closeable {
                     }
                 });
 
+                final ApnsClientHandler apnsClientHandler;
+                {
+                    final String authority = apnsServerAddress.getHostName();
+
+                    final ApnsClientHandler.ApnsClientHandlerBuilder clientHandlerBuilder;
+
+                    if (signingKey != null) {
+                        clientHandlerBuilder = new TokenAuthenticationApnsClientHandler.TokenAuthenticationApnsClientHandlerBuilder()
+                                .signingKey(signingKey)
+                                .tokenExpirationMillis(tokenExpirationMillis)
+                                .authority(authority)
+                                .idlePingIntervalMillis(idlePingIntervalMillis);
+                    } else {
+                        clientHandlerBuilder = new ApnsClientHandler.ApnsClientHandlerBuilder()
+                                .authority(authority)
+                                .idlePingIntervalMillis(idlePingIntervalMillis);
+                    }
+
+                    if (frameLogger != null) {
+                        clientHandlerBuilder.frameLogger(frameLogger);
+                    }
+
+                    apnsClientHandler = clientHandlerBuilder.build();
+
+                    if (gracefulShutdownTimeoutMillis > 0) {
+                        apnsClientHandler.gracefulShutdownTimeoutMillis(gracefulShutdownTimeoutMillis);
+                    }
+                }
+
+                final ChannelPipeline pipeline = channel.pipeline();
+
+                if (proxyHandlerFactory != null) {
+                    pipeline.addFirst(proxyHandlerFactory.createProxyHandler());
+                }
+
                 pipeline.addLast(sslHandler);
-                pipeline.addLast(ConnectionNegotiationErrorHandler.INSTANCE);
+                pipeline.addLast(new FlushConsolidationHandler(FlushConsolidationHandler.DEFAULT_EXPLICIT_FLUSH_AFTER_FLUSHES, true));
+                pipeline.addLast(new IdleStateHandler(idlePingIntervalMillis, 0, 0, TimeUnit.MILLISECONDS));
+                pipeline.addLast(apnsClientHandler);
             }
         });
     }

--- a/pushy/src/main/java/com/eatthepath/pushy/apns/ApnsChannelFactory.java
+++ b/pushy/src/main/java/com/eatthepath/pushy/apns/ApnsChannelFactory.java
@@ -42,8 +42,6 @@ import io.netty.util.concurrent.Future;
 import io.netty.util.concurrent.GenericFutureListener;
 import io.netty.util.concurrent.Promise;
 import io.netty.util.concurrent.PromiseNotifier;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
 
 import java.io.Closeable;
 import java.net.InetSocketAddress;
@@ -70,10 +68,8 @@ class ApnsChannelFactory implements PooledObjectFactory<Channel>, Closeable {
     private static final long MIN_CONNECT_DELAY_SECONDS = 1;
     private static final long MAX_CONNECT_DELAY_SECONDS = 60;
 
-    private static final AttributeKey<Promise<Channel>> CHANNEL_READY_PROMISE_ATTRIBUTE_KEY =
+    static final AttributeKey<Promise<Channel>> CHANNEL_READY_PROMISE_ATTRIBUTE_KEY =
             AttributeKey.valueOf(ApnsChannelFactory.class, "channelReadyPromise");
-
-    private static final Logger log = LoggerFactory.getLogger(ApnsChannelFactory.class);
 
     ApnsChannelFactory(final SslContext sslContext, final ApnsSigningKey signingKey, final long tokenExpirationMillis,
                        final ProxyHandlerFactory proxyHandlerFactory, final int connectTimeoutMillis,
@@ -106,17 +102,6 @@ class ApnsChannelFactory implements PooledObjectFactory<Channel>, Closeable {
             @Override
             protected void initChannel(final SocketChannel channel) {
                 final SslHandler sslHandler = sslContext.newHandler(channel.alloc());
-
-                sslHandler.handshakeFuture().addListener(new GenericFutureListener<Future<Channel>>() {
-                    @Override
-                    public void operationComplete(final Future<Channel> handshakeFuture) {
-                        if (handshakeFuture.isSuccess()) {
-                            channel.attr(CHANNEL_READY_PROMISE_ATTRIBUTE_KEY).get().trySuccess(channel);
-                        } else {
-                            tryFailureAndLogRejectedCause(channel.attr(CHANNEL_READY_PROMISE_ATTRIBUTE_KEY).get(), handshakeFuture.cause());
-                        }
-                    }
-                });
 
                 final ApnsClientHandler apnsClientHandler;
                 {
@@ -203,25 +188,10 @@ class ApnsChannelFactory implements PooledObjectFactory<Channel>, Closeable {
                     @Override
                     public void operationComplete(final ChannelFuture future) {
                         if (!future.isSuccess()) {
-                            // This may seem spurious, but our goal here is to accurately report the cause of
-                            // connection failure; if we just wait for connection closure, we won't be able to
-                            // tell callers anything more specific about what went wrong.
-                            tryFailureAndLogRejectedCause(channelReadyPromise, future.cause());
+                            channelReadyPromise.tryFailure(future.cause());
                         }
                     }
                 });
-
-                connectFuture.channel().closeFuture().addListener(new GenericFutureListener<ChannelFuture> () {
-
-                    @Override
-                    public void operationComplete(final ChannelFuture future) {
-                        // We always want to try to fail the "channel ready" promise if the connection closes; if it has
-                        // already succeeded, this will have no effect.
-                        channelReadyPromise.tryFailure(
-                                new IllegalStateException("Channel closed before HTTP/2 preface completed."));
-                    }
-                });
-
             }
         }, delay, TimeUnit.SECONDS);
 
@@ -250,12 +220,6 @@ class ApnsChannelFactory implements PooledObjectFactory<Channel>, Closeable {
             if (this.hasReleasedSslContext.compareAndSet(false, true)) {
                 ((ReferenceCounted) this.sslContext).release();
             }
-        }
-    }
-
-    private static void tryFailureAndLogRejectedCause(final Promise<?> promise, final Throwable cause) {
-        if (!promise.tryFailure(cause)) {
-            log.warn("Tried to mark promise as \"failed,\" but it was already done.", cause);
         }
     }
 }

--- a/pushy/src/main/java/com/eatthepath/pushy/apns/ApnsClientHandler.java
+++ b/pushy/src/main/java/com/eatthepath/pushy/apns/ApnsClientHandler.java
@@ -28,6 +28,7 @@ import com.google.gson.GsonBuilder;
 import com.eatthepath.pushy.apns.util.DateAsTimeSinceEpochTypeAdapter;
 import io.netty.buffer.ByteBuf;
 import io.netty.buffer.Unpooled;
+import io.netty.channel.Channel;
 import io.netty.channel.ChannelFuture;
 import io.netty.channel.ChannelHandlerContext;
 import io.netty.channel.ChannelPromise;
@@ -38,7 +39,10 @@ import io.netty.handler.codec.http2.*;
 import io.netty.handler.timeout.IdleStateEvent;
 import io.netty.util.AsciiString;
 import io.netty.util.collection.IntObjectHashMap;
-import io.netty.util.concurrent.*;
+import io.netty.util.concurrent.GenericFutureListener;
+import io.netty.util.concurrent.Promise;
+import io.netty.util.concurrent.PromiseCombiner;
+import io.netty.util.concurrent.ScheduledFuture;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -388,6 +392,11 @@ class ApnsClientHandler extends Http2ConnectionHandler implements Http2FrameList
     @Override
     public void onSettingsRead(final ChannelHandlerContext context, final Http2Settings settings) {
         log.debug("Received settings from APNs gateway: {}", settings);
+
+        // Always try to mark the "channel ready" promise as a success after we receive a SETTINGS frame. If it's the
+        // first SETTINGS frame, we know all handshaking and connection setup is done and the channel is ready to use.
+        // If it's a subsequent SETTINGS frame, this will have no effect.
+        getChannelReadyPromise(context.channel()).trySuccess(context.channel());
     }
 
     @Override
@@ -493,5 +502,15 @@ class ApnsClientHandler extends Http2ConnectionHandler implements Http2FrameList
         }
 
         this.unattachedResponsePromisesByStreamId.clear();
+    }
+
+    public void exceptionCaught(final ChannelHandlerContext context, final Throwable cause) {
+        // Always try to fail the "channel ready" promise if we catch an exception; in some cases, these may happen
+        // after a connection has already become ready, in which case the failure attempt will have no effect.
+        getChannelReadyPromise(context.channel()).tryFailure(cause);
+    }
+
+    private Promise<Channel> getChannelReadyPromise(final Channel channel) {
+        return channel.attr(ApnsChannelFactory.CHANNEL_READY_PROMISE_ATTRIBUTE_KEY).get();
     }
 }

--- a/pushy/src/test/java/com/eatthepath/pushy/apns/ApnsClientTest.java
+++ b/pushy/src/test/java/com/eatthepath/pushy/apns/ApnsClientTest.java
@@ -248,8 +248,18 @@ public class ApnsClientTest extends AbstractClientServerTest {
             assertFalse("Clients must not connect to untrusted servers.",
                     sendFuture.isSuccess());
 
+            boolean hasSSLHandshakeException = false;
+            {
+                Throwable cause = sendFuture.cause();
+
+                while (!hasSSLHandshakeException && cause != null) {
+                    hasSSLHandshakeException = (cause instanceof SSLHandshakeException);
+                    cause = cause.getCause();
+                }
+            }
+
             assertTrue("Clients should refuse to connect to untrusted servers due to an SSL handshake failure.",
-                    sendFuture.cause() instanceof SSLHandshakeException);
+                    hasSSLHandshakeException);
         } finally {
             cautiousClient.close().await();
             server.shutdown().await();

